code/datums/json_savefile.dm

```diff
@@ -0,0 +1,71 @@
+/**
+ * A savefile implementation that handles all data using json
+ * also saves it using JSON too, fancy
+ */
+/datum/json_savefile
+	var/path
+	VAR_PRIVATE/list/tree
+	var/auto_save = FALSE
+
+/datum/json_savefile/New(path)
+	src.path = path
+	tree = list()
+	if(fexists(path))
+		load()
+
+/**
+ * Gets an entry from the json tree, with an optional default value.
+ * If no key is specified it throws the entire tree at you instead
+ */
+/datum/json_savefile/proc/get_entry(key, default_value)
+	if(!key)
+		return tree
+	return (key in tree) ? tree[key] : default_value
+
+/// Sets an entry in the tree to the given value
+/datum/json_savefile/proc/set_entry(key, value)
+	tree[key] = value
+	if(auto_save)
+		save()
+
+/// Removes the given key from the tree, if no key is given it clears the entire tree
+/datum/json_savefile/proc/clear(key)
+	if(key)
+		tree -= key
+	else
+		tree.Cut()
+	if(auto_save)
+		save()
+
+/datum/json_savefile/proc/load()
```

Comment:
Same here WRT security. It's theoretically possible for, say, a privileged attacker to read a secured JSON file.

Reply:
vv_edit_var on path to deny editing it

Reply:
also check the admin created datum flag here and return if its set.

also use can proc call to prevent proc calling New() manually.